test/framework/image_preloading.go

```diff
@@ -0,0 +1,219 @@
+	ctrlClient := clusterProxy.GetClient()
+
+	// Create the DaemonSet.
+	_, err := controllerutil.CreateOrPatch(ctx, ctrlClient, daemonSet, daemonSetMutateFn)
+	Expect(err).ToNot(HaveOccurred())
+
+	// Wait for DaemonSet to be available.
+	waitForDaemonSetAvailable(ctx, waitForDaemonSetAvailableInput{Getter: ctrlClient, Daemonset: daemonSet}, time.Minute*3, time.Second*10)
+
+	// List all pods and load images via each found pod.
+	pods := &corev1.PodList{}
+	Expect(ctrlClient.List(
+		ctx,
+		pods,
+		client.InNamespace(daemonSet.Namespace),
+		client.MatchingLabels(daemonSetLabels),
+	)).To(Succeed())
+
+	errs := []error{}
+	for j := range pods.Items {
+		pod := pods.Items[j]
+		Byf("Loading images to node %s via pod %s", pod.Spec.NodeName, klog.KObj(&pod))
+		if err := loadImagesViaPod(ctx, clusterProxy, sourceFile, pod.Namespace, pod.Name, pod.Spec.Containers[0].Name); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	Expect(kerrors.NewAggregate(errs)).ToNot(HaveOccurred())
+
+	// Delete the DaemonSet.
+	Expect(ctrlClient.Delete(ctx, daemonSet)).To(Succeed())
+}
+
+func loadImagesViaPod(ctx context.Context, clusterProxy framework.ClusterProxy, sourceFile, namespace, podName, containerName string) error {
+	// Open source tar file.
+	reader, writer := io.Pipe()
+	file, err := os.Open(filepath.Clean(sourceFile))
+	if err != nil {
+		return err
+	}
+
+	// Use go routine to pipe source file content into then stdin.
+	go func(file *os.File, writer io.WriteCloser) {
+		defer writer.Close()
+		defer file.Close()
+		// Ignoring the error here because the execPod command should fail in case of
+		// failure copying over the data.
+		_, err := io.Copy(writer, file)
+		if err != nil {
+			fmt.Fprintf(ginkgo.GinkgoWriter, "Failed to copy file data to io.Pipe: %v\n", err)
+		}
+	}(file, writer)
+
+	// Load the container images using ctr and delete the file.
+	loadCommand := "ctr -n k8s.io images import -"
+	return execPod(ctx, clusterProxy, namespace, podName, containerName, loadCommand, reader)
+}
+
+// execPod executes a command at a pod.
+// xref: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/exec/exec.go#L123
+func execPod(ctx context.Context, clusterProxy framework.ClusterProxy, namespace, podName, containerName, cmd string, stdin io.Reader) error {
+	var hasStdin bool
+	if stdin != nil {
+		hasStdin = true
+	}
+
+	req := clusterProxy.GetClientSet().CoreV1().RESTClient().Post().
+		Namespace(namespace).
+		Resource("pods").
+		Name(podName).
+		SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Container: containerName,
+			Command:   []string{"/bin/sh", "-c", cmd},
+			Stdin:     hasStdin,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(clusterProxy.GetRESTConfig(), "POST", req.URL())
+	if err != nil {
+		return err
+	}
+
+	var stdout, stderr bytes.Buffer
```

Comment:
Why is this missing the following?
```go
	// WebSocketExecutor must be "GET" method as described in RFC 6455 Sec. 4.1 (page 17).
	websocketExec, err := remotecommand.NewWebSocketExecutor(clusterProxy.GetRESTConfig(), "GET", req.URL().String())
	if err != nil {
		return err
	}
	exec, err = remotecommand.NewFallbackExecutor(websocketExec, exec, httpstream.IsUpgradeFailure)
	if err != nil {
		return err
	}
```

Reply:
because these are not available in the used client-go version.

Reply:
Okay, but it just works without?